### PR TITLE
producers, command line options, and kafka

### PR DESCRIPTION
**Pull request by osheroff**
@vanchi-zendesk

### CODE REVIEW

File: src/main/java/com/zendesk/maxwell/MaxwellConfig.java [hunk 1]
```diff
@@ -57,31 +66,89 @@ public void setInitialPosition(BinlogPosition position) throws IOException {
...
 		}
+
+	}
+
+	public static MaxwellConfig buildConfig(String filename, String [] argv) throws IOException {
+		MaxwellConfig config = new MaxwellConfig();
+
+		config.parseFile(filename);
+		config.parseOptions(argv);
+		config.setDefaults();
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellConfig.java]: Tough to understand precedence here, you would expect defaults to be at the top.

File: src/main/java/com/zendesk/maxwell/MaxwellConfig.java [hunk 2]
```diff
@@ -57,31 +66,89 @@ public void setInitialPosition(BinlogPosition position) throws IOException {
...
+		if ( options.has("host"))
+			this.mysqlHost = (String) options.valueOf("host");
+		if ( options.has("password"))
+			this.mysqlPassword = (String) options.valueOf("password");
+		if ( options.has("user"))
+			this.mysqlUser = (String) options.valueOf("user");
+		if ( options.has("port"))
+			this.mysqlPort = Integer.valueOf((String) options.valueOf("port"));
+		if ( options.has("producer"))
+			this.producer = (String) options.valueOf("producer");
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellConfig.java]: Maybe rename producer -> producerType. We also have a getProducer() in here which actually creates the producer object.

File: src/main/java/com/zendesk/maxwell/producer/MaxwellKafkaProducer.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	public KafkaCallback(MaxwellAbstractRowsEvent e, MaxwellConfig c) {
+		this.config = c;
+		this.event = e;
+	}
+
+	@Override
+	public void onCompletion(RecordMetadata md, Exception e) {
+		if ( e != null ) {
+            e.printStackTrace();
+		} else {
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/producer/MaxwellKafkaProducer.java]: nitpic: indentation

File: src/main/java/com/zendesk/maxwell/producer/AbstractProducer.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+public abstract class AbstractProducer {
+	protected final MaxwellConfig config;
+
+	public AbstractProducer(MaxwellConfig config) {
+		this.config = config;
+	}
+	abstract public void push(MaxwellAbstractRowsEvent e) throws Exception;
+
+	public void onComplete(MaxwellAbstractRowsEvent e) {
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/producer/AbstractProducer.java]: Is this called anywhere?

Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/producer/AbstractProducer.java]: in the subclasses.  but I don't think it's needed in the base class.

**Comment by vanchi-zendesk** [comment]
:+1:

File: src/main/java/com/zendesk/maxwell/producer/AbstractProducer.java (hunk 4 above)
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/producer/AbstractProducer.java]: scratch that; it's used in the FileClass.

**Comment by osheroff** [comment]
updated a bit.

**Comment by vanchi-zendesk** [comment]
:+1:

**Pull request merged**